### PR TITLE
Non-blocking case locks

### PR DESCRIPTION
**Pull request by millerdev**
Case locks acquired in the form processor's case db cache (and by a few management commands) are now non-blocking. This should remove the possibility of deadlock when multiple requests lock a set of cases with different lock acquisition ordering. Note that this deadlock scenario would result in the deadlocked requests waiting for lock expiration before continuing with form processing with expired locks, which may be worse than not using locks at all.

Example:
- request 1 locks case A
- request 2 locks case B
- request 1 locks case B (block)
- request 2 locks case A (block, deadlock until timeout)
- lock A expires
- lock B expires
- request 1 locks case B (request 2 thinks it still has a valid lock on case B)
- request 2 locks case A (request 1 thinks it still has a valid lock on case A)
- request 1 finishes form processing without lock on case A
- request 2 finishes form processing without lock on case B

This is just one possible, and probably worst case, scenario. Slight changes in timing of lock expiration and acquisition may result in one of the requests completing with valid locks. However, it is not possible (in the case of deadlock) for both requests to complete with valid locks.

Would it make sense to write form submissions to a queue rather than processing synchronously while the phone waits? That should allow a faster response and may reduce the number of times a phone needs to retry sending a submission.

Probably best reviewed by commit 🐠 

@snopoke @dannyroberts

### CODE REVIEW

**Comment by millerdev** [comment]
Test failures are real. One is an easy fix and the other not so easy. Discovered that we have some tests that block until lock timeout and then go on to pass 🙃

File: corehq/ex-submodules/dimagi/utils/couch/__init__.py [hunk 1]
```diff
@@ -222,18 +232,20 @@ def get_locked_obj(cls, *args, **kwargs):
...
+            obj_lock = cls.get_obj_lock(obj)
+            obj_lock = acquire_lock(obj_lock, degrade_gracefully, blocking=block)
+            if block:
                 # Refresh the object in case another thread has updated it
                 obj = cls.get_latest_obj(obj)
-                release_lock(lock, degrade_gracefully)
-                return LockManager(obj, obj_lock)
+            return LockManager(obj, obj_lock)
+        finally:
+            release_lock(lock, degrade_gracefully)
```
Comment (snopoke) [inline corehq/ex-submodules/dimagi/utils/couch/__init__.py]: this code always confuses me, too many locks and alternate code paths!

**Review by snopoke — APPROVED** [review]
Hard to think through all the ramifications of this but I think I agree it's a better situation.

**Comment by snopoke** [comment]
> Would it make sense to write form submissions to a queue rather than processing synchronously while the phone waits? That should allow a faster response and may reduce the number of times a phone needs to retry sending a submission.

This option has come up a few times in the past but we've never gotten it to design stage. There are some challenges in terms of making sure all forms from a device are processed prior to allowing a sync. I'd be keen for us to flesh it out more though to see how much effort it would be.

File: corehq/ex-submodules/dimagi/utils/couch/__init__.py (hunk 1 above)
Comment (millerdev) [reply corehq/ex-submodules/dimagi/utils/couch/__init__.py]: I agree!

**Comment by snopoke** [comment]
Closing old PRs. Reopen if necessary.

**Pull request closed without merging**